File: 2023/challenge_04/challenge_04.py
import os
import sys

def read_file():
  path = os.path.dirname(__file__)
  with open(os.path.join(path, 'challenge_04.txt'), encoding='UTF-8') as file:
    return file.read().splitlines()

def run():
  try:
    param = int(sys.argv[1])
  except:
    print('⭕ Eror: No se proporciono el parametro')
    sys.exit(1)

  file_names = read_file()
  count = 0

  for name in file_names:
    name, checksum = name.split('-')
    correct_checksum = ''.join(list(filter(lambda x: name.count(x) == 1, name)))

    if correct_checksum == checksum:
      count += 1
      if count == param:
        return correct_checksum

  return '❗ No se encontro el archivo'

if __name__ == '__main__':
  print(run())
  

File: 2023/challenge_04/challenge_04.js
const fs = require('node:fs/promises')
const path = require('node:path')

async function read_file() {
  const list = await fs.readFile(path.join(__dirname, 'challenge_04.txt'), 'utf8');
  return list.split('\n');
}

async function run() {
  const listFiles = await read_file();
  const arg = Number(process.argv[2] ?? 1);
  let count = 0;

  for (const file of listFiles) {
    const [filename, checksum] = file.trim().split('-');
    const correctChecksum = filename.split('').reduce((acc, letter, index, arr) => {
      if (arr.filter(elem => elem === letter).length > 1) return acc;
      return acc += letter;
    }, '');

    if (correctChecksum === checksum) {
      count++;
      if (count === arg) return checksum;
    }
  }

  return '❗ No se encontro el archivo';
}

run().then(console.log)